### PR TITLE
feat: external tfs service endpoint for azure devops / tfs

### PR DESCRIPTION
**Pull request by arnonearth**
## All Submissions:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.
* [x] All new and existing tests passed.
* [x] My code follows the code style of this project.
* [x] I ran lint checks locally prior to submission.
* [x] Have you checked to ensure there aren't other open PRs for the same update/change?

## What about the current behavior has changed?

Issue Number: #339

## Does this introduce a change to `go.mod`, `go.sum` or `vendor/`?

- [ ] Yes
- [x] No

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Other information

I have run unit tests, acceptance tests, and format/lint checks accordingly.

I used a customized version of the provider at the moment so comments and feedbacks are always welcome.

### A screenshot of unit test result. (local run)
unittest-result

### A screenshot of acceptance test result. (tested on my personal azure devops organization)
acctest-result

### A screenshot of azure pipelines based-on azure-pipelines.yml & ci.yml template without tests for format, lint, and build checks.
fmt-lint-check

### CODE REVIEW

**Comment by arnonearth** [comment]
@microsoft-github-policy-service agree [company="Axons"]

**Comment by arnonearth** [comment]
@microsoft-github-policy-service agree company="Axons"

File: website/docs/r/serviceendpoint_externaltfs.html.markdown [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+## Argument Reference
+
+The following arguments are supported:
+
+- `project_id` - (Required) The ID of the project.
+- `service_endpoint_name` - (Required) The Service Endpoint name.
+- `url` - (Required) Azure DevOps Organization or TFS Project Collection Url.
+- `description` - (Optional) The Service Endpoint description. Defaults to `Managed by Terraform`.
+- `auth_personal` - (Optional) An `auth_personal` block as documented below. Allows connecting using a personal access token.
```
Comment (xuzhang3) [inline website/docs/r/serviceendpoint_externaltfs.html.markdown]: `auth_personal` is a required parameter not optional

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_externaltfs.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+				Type:         schema.TypeString,
+				Required:     true,
+				DefaultFunc:  schema.EnvDefaultFunc("AZDO_PERSONAL_ACCESS_TOKEN", nil),
+				Description:  "Personal access tokens are applicable only for connections targeting Azure DevOps organization or TFS 2017 (and higher)",
+				Sensitive:    true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+		},
+	}
+	patHashKey, patHashSchema := tfhelper.GenerateSecreteMemoSchema(personalAccessTokenExternalTFS)
```
Comment (xuzhang3) [inline azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_externaltfs.go]: hash functions has be deprecated, `patHashKey` and  `patHashSchema` can be removed.

**Review by xuzhang3 — COMMENTED** [review]
@arnonearth  Thanks for your PR. A few changes requested, others looks good to me

File: website/docs/r/serviceendpoint_externaltfs.html.markdown (hunk 1 above)
Comment (arnonearth) [reply website/docs/r/serviceendpoint_externaltfs.html.markdown]: Updated.

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_externaltfs.go (hunk 2 above)
Comment (arnonearth) [reply azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_externaltfs.go]: Code-related to the deprecated functions were removed.

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_externaltfs.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+func flattenServiceEndpointExternalTFS(
+	d *schema.ResourceData,
+	serviceEndpoint *serviceendpoint.ServiceEndpoint,
+	projectID *uuid.UUID,
+) {
+	doBaseFlattening(d, serviceEndpoint, projectID)
+
+	d.Get("")
```
Comment (xuzhang3) [inline azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_externaltfs.go]: unused code

**Comment by xuzhang3** [comment]
```log
==> Checking that code complies with gofmt requirements...
=== RUN   TestAccServiceEndpointExternalTFS_RequiresImportErrorStep
=== PAUSE TestAccServiceEndpointExternalTFS_RequiresImportErrorStep
=== CONT  TestAccServiceEndpointExternalTFS_PersonalTokenBasic
=== CONT  TestAccServiceEndpointExternalTFS_CreateAndUpdate
=== CONT  TestAccServiceEndpointExternalTFS_RequiresImportErrorStep
=== CONT  TestAccServiceEndpointExternalTFS_PersonalTokenUpdate
--- PASS: TestAccServiceEndpointExternalTFS_PersonalTokenBasic (66.42s)
--- PASS: TestAccServiceEndpointExternalTFS_RequiresImportErrorStep (69.83s)
--- PASS: TestAccServiceEndpointExternalTFS_PersonalTokenUpdate (97.70s)
--- PASS: TestAccServiceEndpointExternalTFS_CreateAndUpdate (98.83s)
PASS
ok      github.com/microsoft/terraform-provider-azuredevops/azuredevops/internal/acceptancetests        114.948s

```

**Review by xuzhang3 — APPROVED** [review]

**Comment by xuzhang3** [comment]
LGTM

**Pull request merged**